File: 3152-special-array-ii/3152-special-array-ii.go
func isArraySpecial(nums []int, queries [][]int) []bool {
    n := len(nums)
    arr := make([]int, n)

    for i := 0; i <n-1; i++ {
        if (nums[i]%2 == nums[i+1]%2) {
            arr[i+1] = 1
        }
        arr[i+1] += arr[i]
    }

    res := []bool{}
    for _, query := range queries {
        res = append(res, arr[query[1]] - arr[query[0]] == 0)
    }
    return res
}

File: 3152-special-array-ii/3152-special-array-ii.rs
impl Solution {
    pub fn is_array_special(nums: Vec<i32>, queries: Vec<Vec<i32>>) -> Vec<bool> {
        let n : usize = nums.len();
        let mut arr : Vec<i32> = vec![0; n];

        for i in 0..n-1 {
            if nums[i]%2 == nums[i+1]%2 {
                arr[i+1] = 1;
            }
            arr[i+1] += arr[i];
        }
        let mut res : Vec<bool> = Vec::with_capacity(queries.len());

        for query in queries {
            if arr[query[1] as usize] - arr[query[0] as usize] != 0 {
                res.push(false);
            } else {
                res.push(true);
            }
        }

        res
    }
}